test/alternator/test_gsi.py

```diff
@@ -163,9 +163,25 @@ def test_gsi_strong_consistency(test_table_gsi_1):
 
 # Test that setting an indexed string column to an empty string is illegal,
 # since keys cannot contain empty strings
-def test_gsi_empty_value(test_table_gsi_2):
+def test_gsi_empty_value(test_table_gsi_3):
     with pytest.raises(ClientError, match='ValidationException.*empty'):
-        test_table_gsi_2.put_item(Item={'p': random_string(), 'x': ''})
+        test_table_gsi_3.put_item(Item={'p': random_string(), 'a': ''})
+    with pytest.raises(ClientError, match='ValidationException.*empty'):
+        test_table_gsi_3.put_item(Item={'p': random_string(), 'b': ''})
+
+# Dynamodb supports special way of setting NULL value.
+# It's different than non existing value (see also test_update_expected_1_not_null).
+def test_gsi_null_value(test_table_gsi_3):
+    with pytest.raises(ClientError, match='ValidationException.*NULL'):
+        test_table_gsi_3.put_item(Item={'p': random_string(), 'a': None})
+    with pytest.raises(ClientError, match='ValidationException.*NULL'):
+        test_table_gsi_3.put_item(Item={'p': random_string(), 'b': None})
+    # index keys can't be empty nor NULL but can be non existent
+    test_table_gsi_3.put_item(Item={'p': random_string()})
+    test_table_gsi_3.put_item(Item={'p': random_string(), 'a': random_string()})
+    test_table_gsi_3.put_item(Item={'p': random_string(), 'b': random_string()})
+    time.sleep(1) # since GSI index read is eventually consistent some compromise here
```

Comment:
It's not a good idea to sleep 1 second. 1 second may be (much) too much, and in other cases might not be enough.
But you're right that the sleep loop in assert_index_scan is not good enough because if the view write gets delayed, we can get an empty result set immediately which doesn't say much.
So I don't know what to suggest instead.

In any case, I would separate this last part of the test ("index keys can't be empty nor NULL but can be non existent") into its own separate test. Then you'll notice it's a duplicate of the existing test test_gsi_missing_attribute - which as you can see also isn't perfect but relies that if we have a bug it will at least **sometimes** show through.

Reply:
ok, cool, if this part is already tested I'll remove it